bin/octocatalog-diff

```diff
@@ -47,46 +67,84 @@ def get_hosts_from_puppetdb():
 
 
 def diff(host):
-    """Run octocatalog-diff for a particular hostname"""
+    """Run octocatalog-diff for a particular hostname, and return whether the
+    run succeeded or failed and any output"""
     process = subprocess.run(
-        (
-            'octocatalog-diff',
-            # Add a --debug flag to get much more verbose output
-            '-n', host,
-            '--enc-override',
-            'environment=production,parameters::dummy_secrets=true',
-            '--display-detail-add',
-            # Ignore changes to puppetserver settings, since these are
-            # dependent on the path that octocatalog-diff sets and thus always
-            # differ
-            '--ignore', "'Ini_setting[puppet.conf/master/storeconfigs]'",
-            '--ignore', "'Ini_setting[puppet.conf/master/storeconfigs_backend]'",
-            '--ignore', "'Ini_setting[puppetdbserver_urls]'",
-            '--ignore', "'Ini_setting[soft_write_failure]'",
-            '--ignore', "'File[/tmp/*/routes.yaml]'",
-            # Ignore changes to SSH keys, since these change when any machines
-            # are reprovisioned, etc. and change frequently
-            '--ignore', "'Sshkey[*]'",
+        ' '.join(
+            (
+                'octocatalog-diff',
```

Comment:
You can (in fact should maybe) pass a list of arguments to `subprocess.run`.

Reply:
So I initially used that, but that kept giving junk diffs for `lightning.ocf.berkeley.edu` where it would still include things like `Ini_setting[puppet.conf/master/storeconfigs]` that are meant to be ignored. I couldn't reproduce the same issue when running the command locally in a shell, and switching this to use a shell also made it work (see the comment on lines 99-101). I'm not a fan of using this either where not having a shell would be better, but I also haven't found a better workaround or dug deep enough into what differs between `octocatalog-diff` when it has a shell versus when it does not to figure this one out, so I'm inclined to leave it for the time being and add a TODO or similar note about switching this in the future if a workaround is found.

Reply:
You can use shell=True and pass a list (though the Python docs are not very clear here). Unless you mean that passing a string vs a list made a difference?

Reply:
>If shell is `True`, it is recommended to pass args as a string rather than as a sequence.
and later
> That is to say, [when `shell=True`,] `Popen` does the equivalent of `Popen(['/bin/sh', '-c', args[0], args[1], ...])`

From the python3 docs on `subprocess.Popen`. So, passing a list will not work.

Reply:
Ah I missed that part, yeah this is fine then.